capstone/src/main/java/com/google/sps/servlets/LessonHandlerServlet.java

```diff
@@ -64,15 +65,44 @@ public void doPost(HttpServletRequest request, HttpServletResponse response) thr
 
     public List<Lesson> getLessons(Room room) {
         List<Lesson> lessons = new ArrayList<>();
+        List<Key> nullKeys = new ArrayList<>();
     
         try {
             for(Key key : room.getAllLessons()) {
-                lessons.add(DatabaseService.getLesson(key));
+                Boolean status = checkLessonStatus(key);
+
+                if (status) {
+                    lessons.add(DatabaseService.getLesson(key));
+                } else {
+                    nullKeys.add(key);
+                }
             }
         }
         catch (EntityNotFoundException e) {
             System.err.println("Lesson entities don't exist.");
         }
+        removeDeletedLessons(room, nullKeys);
         return lessons;
     }
+
+    // Checks to see if the lesson still exists in datastore
+    public Boolean checkLessonStatus(Key key) {
+        DatastoreService datastore = DatastoreServiceFactory.getDatastoreService();
+        Query query = new Query(Lesson.LESSON_ENTITY_NAME).addFilter(Entity.KEY_RESERVED_PROPERTY, FilterOperator.EQUAL, key);
+        Entity result = datastore.prepare(query).asSingleEntity();
+
+        if (result == null) {
+            return false;
+        }
+        return true;
+    }
+
+    public void removeDeletedLessons(Room room, List<Key> keys) {
+        List<Key> lessons = room.getAllLessons();
+        for (Key key : keys) {
+            lessons.remove(key);
+        }
+        room.getRoomEntity().setProperty(Room.LESSONS_PROPERTY_KEY, lessons);
+        DatabaseService.save(room.getRoomEntity());
+    }
```

Comment:
This kind of cleanup (removing the lesson key from the parent room) should be done ahead of time when the lesson is removed (part of the "appropriate cleanup" relevant to lessons mentioned above), not upon requesting lessons. (This means this file should have basically no diff other than moving what was already existing for the getLessons to DatabaseService)